### PR TITLE
 Resolve degenerate performance in `create_structs_data`

### PR DESCRIPTION
**Pull request by SurajAralihalli**
Resolves issue [#14716](https://github.com/rapidsai/cudf/issues/14716)

- Eliminated unnecessary recursive self-calls in the `superimpose_nulls_no_sanitize` function, addressing performance issues in `make_structs_column`.
- Introduced `STRUCT_CREATION_NVBENCH` to assess the performance of the `create_structs_data` function.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by SurajAralihalli** [comment]
Example structure definition: `STRUCT<STRUCT<INT32>> : [ {{Null}}, {Null}, Null, {{3}} ]`

In the code:
```
P {
  type = STRUCT
  null_mask = [1, 1, 0, 1]
  null_count = 1
  children =
    C_1 {
      type = STRUCT
      null_mask = [1, 0, 0, 1]
      null_count = 2
      children =
        C_2 {
          type = INT32
          data =       [X, X, X, 3]
          null_mask  = [0, 0, 0, 1]
          null_count = 3
        }
    }
}
```


The `make_structs_column` function applies the null mask of the parent (P) to its immediate children (C_1) at line https://github.com/rapidsai/cudf/blob/07103355fea0fb3fd0e1115019bbac7d65bb132f/cpp/src/structs/utilities.cpp#L244. 
The `superimpose_nulls_no_sanitize` function recursively applies a correct null mask to all nested children (C_2) at line https://github.com/rapidsai/cudf/blob/07103355fea0fb3fd0e1115019bbac7d65bb132f/cpp/src/structs/utilities.cpp#L266. 
This function uses `make_structs_column` to reconstruct the parent column (P) which applies C_1's null mask to its children (C_2) at line https://github.com/rapidsai/cudf/blob/07103355fea0fb3fd0e1115019bbac7d65bb132f/cpp/src/structs/utilities.cpp#L382 However, C_2's null_mask is already correct.

This redundancy can be eliminated by removing the code at line [266](https://github.com/rapidsai/cudf/blob/07103355fea0fb3fd0e1115019bbac7d65bb132f/cpp/src/structs/utilities.cpp#L266).

**Comment by SurajAralihalli** [comment]
Before:
Screenshot 2024-01-15 at 10 59 16 PM

After:
Screenshot 2024-01-15 at 11 01 23 PM

Thank you @karthikeyann for sharing scope based NVTX ranges https://github.com/rapidsai/cudf/pull/11864#issuecomment-1271514499. This utility greatly helped in identifying the differences.

**Comment by vuule** [comment]
How come there were 130k calls to `superimpose_nulls_no_sanitize`? It's more than even quadratic redundancy (not a thing, but hopefully you know what I meant :)) would cause.

I think this change should include a new benchmark.

File: cpp/benchmarks/CMakeLists.txt [hunk 1]
```diff
@@ -183,6 +183,11 @@ ConfigureNVBench(
   sort/sort_lists.cpp sort/sort_structs.cpp
 )
 
+# ##################################################################################################
+# * create benchmark
+# --------------------------------------------------------------------------------
+ConfigureNVBench(CREATE_NVBENCH create/create_structs.cpp)
```
Comment (ttnghia) [inline cpp/benchmarks/CMakeLists.txt:189]: Let's call something else like `STRUCT_CREATION_NVBENCH`. The prefix `CREATE_` is not clear.

Comment (ttnghia) [reply cpp/benchmarks/CMakeLists.txt:189]: Also please post the comparison of benchmark results. FYI, the comparison is done by
```
python path_to_nvbench/nvbench_compare.py ./before.json after.json
```
where the json files generated by
```
./BENCH_NAME_NVBENCH --json file.json
```

**Comment by SurajAralihalli** [comment]
> How come there were 130k calls to `superimpose_nulls_no_sanitize`? It's more than even quadratic redundancy (not a thing, but hopefully you know what I meant :)) would cause.
> 
> I think this change should include a new benchmark.

Absolutely @vuule! It would be interesting to analyze the complexity relationship between calls to `make_structs_column` and `superimpose_nulls_no_sanitize` in terms of depth. Also, I've created `create_structs`  benchmark. 

`./CREATE_NVBENCH -b create_structs --axis NumRows=[10] --axis Depth=[16] --axis Nulls=[0]`

Before:
```
| NumRows | Depth | Nulls | Samples | CPU Time | Noise | GPU Time | Noise |
|---------|-------|-------|---------|----------|-------|----------|-------|
|      10 |    16 |     0 |      7x |  2.367 s | 0.22% |  2.367 s | 0.22% |
```

After:
```
| NumRows | Depth | Nulls | Samples | CPU Time | Noise | GPU Time | Noise |
|---------|-------|-------|---------|----------|-------|----------|-------|
|      10 |    16 |     0 |   2960x | 2.691 ms | 9.79% | 2.685 ms | 9.79% |
```

File: cpp/benchmarks/CMakeLists.txt (hunk 1 above)
Comment (SurajAralihalli) [reply cpp/benchmarks/CMakeLists.txt]: Thanks @ttnghia, I have updated the PR.

**Comment by SurajAralihalli** [comment]
```
## [0] NVIDIA RTX A5000

|  NumRows  |  Depth  |  Nulls  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |            Diff |   %Diff |  Status  |
|-----------|---------|---------|------------|-------------|------------|-------------|-----------------|---------|----------|
|   2^10    |    1    |    0    |  42.087 us |       7.79% |  39.435 us |       6.45% |       -2.651 us |  -6.30% |   PASS   |
|   2^18    |    1    |    0    |   7.681 ms |       3.37% |   7.582 ms |       2.55% |      -98.575 us |  -1.28% |   PASS   |
|   2^26    |    1    |    0    |    2.346 s |       0.23% |    2.357 s |       0.55% |       11.391 ms |   0.49% |   FAIL   |
|   2^10    |    8    |    0    |   8.984 ms |       1.35% | 824.409 us |       2.86% |    -8159.408 us | -90.82% |   FAIL   |
|   2^18    |    8    |    0    |  47.787 ms |       1.30% |  38.508 ms |       1.60% |    -9278.895 us | -19.42% |   FAIL   |
|   2^26    |    8    |    0    |   10.222 s |        inf% |   10.222 s |        inf% |       29.785 us |   0.00% |   PASS   |
|   2^10    |   16    |    0    |    2.455 s |       0.58% |   3.206 ms |       1.87% | -2452204.726 us | -99.87% |   FAIL   |
|   2^18    |   16    |    0    |    2.779 s |       0.23% |  75.308 ms |       1.82% | -2703843.785 us | -97.29% |   FAIL   |
|   2^26    |   16    |    0    |   26.935 s |        inf% |   19.043 s |        inf% | -7891943.359 us | -29.30% |   PASS   |
|   2^10    |    1    |    1    |  56.118 us |       5.24% |  56.982 us |       6.84% |        0.864 us |   1.54% |   PASS   |
|   2^18    |    1    |    1    |   7.839 ms |       1.51% |   7.985 ms |       3.48% |      146.030 us |   1.86% |   FAIL   |
|   2^26    |    1    |    1    |    2.447 s |       1.11% |    2.434 s |       0.48% |   -12658.447 us |  -0.52% |   FAIL   |
|   2^10    |    8    |    1    |   9.043 ms |       1.18% | 837.039 us |       2.92% |    -8206.161 us | -90.74% |   FAIL   |
|   2^18    |    8    |    1    |  48.371 ms |       1.75% |  38.803 ms |       1.66% |    -9568.242 us | -19.78% |   FAIL   |
|   2^26    |    8    |    1    |   10.312 s |        inf% |   10.285 s |        inf% |   -26727.539 us |  -0.26% |   PASS   |
|   2^10    |   16    |    1    |    2.411 s |       0.27% |   3.235 ms |       1.79% | -2407859.781 us | -99.87% |   FAIL   |
|   2^18    |   16    |    1    |    2.762 s |       0.40% |  75.156 ms |       0.44% | -2686969.612 us | -97.28% |   FAIL   |
|   2^26    |   16    |    1    |   27.152 s |        inf% |   19.126 s |        inf% | -8025222.656 us | -29.56% |   PASS   |
```

File: cpp/benchmarks/create/create_structs.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+{
+  state.exec(nvbench::exec_tag::sync,
+             [&](nvbench::launch& launch) { auto const table_ptr = create_structs_data(state); });
+}
+
+NVBENCH_BENCH(nvbench_create_structs)
+  .set_name("create_structs")
+  .add_int64_power_of_two_axis("NumRows", {10, 18, 26})
+  .add_int64_axis("Depth", {1, 8, 16})
+  .add_int64_axis("Nulls", {0, 1});
```
Comment (ttnghia) [inline cpp/benchmarks/create/create_structs.cpp:31]: Basically this does nothing with the parameters. So the performance is constant in all conditions.

File: cpp/benchmarks/CMakeLists.txt [hunk 3]
```diff
@@ -183,6 +183,11 @@ ConfigureNVBench(
   sort/sort_lists.cpp sort/sort_structs.cpp
 )
 
+# ##################################################################################################
+# * create benchmark
+# --------------------------------------------------------------------------------
+ConfigureNVBench(STRUCT_CREATION_NVBENCH create/create_structs.cpp)
```
Comment (davidwendt) [inline cpp/benchmarks/CMakeLists.txt:189]:
```suggestion
ConfigureNVBench(STRUCT_CREATION_NVBENCH structs/create_structs.cpp)
```
I'd prefer this to be in a `structs` sub-directory at this time.

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/benchmarks/CMakeLists.txt (hunk 3 above)
Comment (SurajAralihalli) [reply cpp/benchmarks/CMakeLists.txt]: done!

**Review by ttnghia — APPROVED** [review]

File: cpp/benchmarks/CMakeLists.txt [hunk 4]
```diff
@@ -183,6 +183,11 @@ ConfigureNVBench(
   sort/sort_lists.cpp sort/sort_structs.cpp
 )
 
+# ##################################################################################################
+# * create benchmark
```
Comment (ttnghia) [inline cpp/benchmarks/CMakeLists.txt:187]: Wait, I missed this. Please update comment here. Also, the new line below should be moved closer to the other struct benchmarks.

Comment (SurajAralihalli) [reply cpp/benchmarks/CMakeLists.txt:187]: Thanks for pointing it out, I missed it too.

Comment (SurajAralihalli) [reply cpp/benchmarks/CMakeLists.txt]: Other struct related benchmarks are in different sub directories eg: sort, groupby

Comment (SurajAralihalli) [reply cpp/benchmarks/CMakeLists.txt]: Resolving this conversation. Please let me know if other any changes are required, thanks!

**Review by davidwendt — APPROVED** [review]

**Pull request merged**